### PR TITLE
erc721 tester

### PR DESCRIPTION
**Pull request by lzxm160**
add erc721 tester：
1、start a single IoTeX server;
2、create two accounts;
3、create a token contract;
4、transfer the token between two accounts;
5、compare the actual sum of tokens with the amount of creation

### CODE REVIEW

File: tools/erc721tester/assetcontract/assetcontract.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
+package assetcontract
```
Comment (raullenchai) [inline tools/erc721tester/assetcontract/assetcontract.go]: copyright header, same below

File: tools/erc721tester/assetcontract/assetcontract.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	"github.com/iotexproject/iotex-core/config"
+	"github.com/iotexproject/iotex-core/tools/erc721tester/blockchain"
+)
+
+const (
+	// ChainIP is the ip address of iotex api endpoint
+	chainIP = "localhost"
+)
+
+// StartContracts deploys and starts fp token smart contract and stable token smart contract
```
Comment (raullenchai) [inline tools/erc721tester/assetcontract/assetcontract.go]: ` fp token smart contract and stable token smart contract` is not true

File: tools/erc721tester/erc721tester.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2019 IoTeX
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package main
+
+import (
+	"context"
```
Comment (raullenchai) [inline tools/erc721tester/erc721tester.go]: group

File: tools/erc721tester/erc721tester.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	cfg.Genesis.ActionGasLimit = 10000000
+	cfg.Genesis.BlockInterval = 2 * time.Second
+	cfg.ActPool.MinGasPriceStr = big.NewInt(0).String()
+	itxsvr, err := itx.NewServer(cfg)
+	if err != nil {
+		log.L().Fatal("Failed to start itxServer.", zap.Error(err))
+	}
+	go itx.StartServer(ctx, itxsvr, probe.New(7799), cfg)
+
+	time.Sleep(2 * time.Second)
```
Comment (raullenchai) [inline tools/erc721tester/erc721tester.go]: waiuntil

File: tools/erc721tester/erc721tester.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	}
+	log.L().Info("Creditor's asset balance: ", zap.Int64("balance", creditorBalance))
+
+
+	//// Transfer erc721 token
+	if _, err := erc721Token.Transfer(erc721Token.Address(), creditorAddr, creditorPriv, debtorAddr, tokenID); err != nil {
+		log.L().Fatal("Failed to transfer 1 token from creditor to debtor", zap.Error(err))
+	}
+
+	debtorBalance, err := erc721Token.ReadValue(erc721Token.Address(), "70a08231", debtorAddr)
```
Comment (raullenchai) [inline tools/erc721tester/erc721tester.go]: don't you have to wait until the transfer tx gets into a block? why can this pass without wait?

File: tools/erc721tester/blockchain/erc721_token.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
...150600090505b85518110156119305785818151811015156118e957fe5b90602001015160f860020a900460f860020a02838380600101945081518110151561191057fe5b906020010190600160f860020a031916908160001a9053506001016118d2565b5060005b845181101561199257848181518110151561194b57fe5b90602001015160f860020a900460f860020a02838380600101945081518110151561197257fe5b906020010190600160f860020a031916908160001a905350600101611934565b5090979650505050505050565b600160a060020a03821615156119b457600080fd5b6119be828261121a565b6040518190600160a060020a038416906000907fddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef908290a45050565b818101828110156108b957fe5b6040805160c081018252600080825260208201819052918101829052606081018290526080810182905260a081019190915290565b81548183558181111561083e5760008381526020902061083e918101908301611ade565b828054600181600116156101000203166002900490600052602060002090601f016020900481019282601f10611aa157805160ff1916838001178555611ace565b82800160010185558215611ace579182015b82811115611ace578251825591602001919060010190611ab3565b50611ada929150611ade565b5090565b61063491905b80821115611ada5760008155600101611ae45600a165627a7a72305820eb6ef4caf7e8dfb40d659802d0d7106ab4b338205d58c4b1d0596c47c740a77e0029"
```
Comment (raullenchai) [inline tools/erc721tester/blockchain/erc721_token.go]: perhaps include the solidity source code in a seperate file, named erc721.sol

File: tools/erc721tester/blockchain/erc721_token.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+func (f *erc721Token) CreateToken(tokenid, creditor string) (string, error) {
+	TokenID,ok:=big.NewInt(0).SetString(tokenid,10)
+	if !ok {
+		return "", errors.Errorf("invalid tokenid = %s", tokenid)
+	}
+	addrCreditor, err := address.FromString(creditor)
+	if err != nil {
+		return "", errors.Errorf("invalid creditor address = %s", creditor)
+	}
+	h,err:=f.RunAsOwner().Call("fdb05e85",addrCreditor.Bytes(),TokenID.Bytes())
```
Comment (raullenchai) [inline tools/erc721tester/blockchain/erc721_token.go]: all these method selector should be defined.

File: tools/erc721tester/blockchain/erc721_token.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2019 IoTeX
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package blockchain
+
+import (
+	"github.com/pkg/errors"
```
Comment (raullenchai) [inline tools/erc721tester/blockchain/erc721_token.go]: group

File: tools/erc721tester/blockchain/contract.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
+// Copyright (c) 2019 IoTeX
```
Comment (raullenchai) [inline tools/erc721tester/blockchain/contract.go]: why need this one? it looks the same as executiontester/blockchain/contract.go

File: tools/erc721tester/erc721tester.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	creditorBalance, err := erc721Token.ReadValue(erc721Token.Address(), "70a08231", creditorAddr)
+	if err != nil {
+		log.L().Fatal("Failed to get creditor's asset balance.", zap.Error(err))
+	}
+	log.L().Info("Creditor's asset balance: ", zap.Int64("balance", creditorBalance))
+
+
+	//// Transfer erc721 token
```
Comment (lizhefeng) [inline tools/erc721tester/erc721tester.go]: Two slashes are enough

File: tools/erc721tester/erc721tester.go (hunk 5 above)
Comment (lizhefeng) [reply tools/erc721tester/erc721tester.go]: There is a backoff retry CheckCallResult inside the transfer implementation.

File: tools/erc721tester/erc721tester.go (hunk 10 above)
Comment (raullenchai) [reply tools/erc721tester/erc721tester.go]: +1

File: tools/erc721tester/erc721tester.go (hunk 5 above)
Comment (raullenchai) [reply tools/erc721tester/erc721tester.go]: ok

File: tools/erc721tester/assetcontract/assetcontract.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+			return t[:8] + id[:8]
+		}
+	}
+}
+
+func deployContract(code, endpoint string, args ...[]byte) (string, error) {
+	// deploy the contract
+	contract := executiontester.NewContract(endpoint)
+
+	setE:=contract.SetExecutor
```
Comment (raullenchai) [inline tools/erc721tester/assetcontract/assetcontract.go]: make lint

File: tools/erc721tester/assetcontract/assetcontract.go [hunk 12]
```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2019 IoTeX
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package assetcontract
```
Comment (lizhefeng) [inline tools/erc721tester/assetcontract/assetcontract.go]: Please reuse "assetcontract.go" in executiontester/assetcontract. See my specific suggestions below.

File: tools/erc721tester/assetcontract/assetcontract.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/iotexproject/iotex-core/tools/erc721tester/blockchain"
+	executiontester "github.com/iotexproject/iotex-core/tools/executiontester/blockchain"
+)
+
+const (
+	// ChainIP is the ip address of iotex api endpoint
+	chainIP = "localhost"
+)
+// StartContracts deploys and starts erc721 token smart contract
+func StartContracts(cfg config.Config) (blockchain.Erc721Token,error) {
```
Comment (lizhefeng) [inline tools/erc721tester/assetcontract/assetcontract.go]: You can upgrade 
```
StartContracts(cfg config.Config) (blockchain.FpToken, blockchain.StableToken, error)
```
to 
```
StartContracts(cfg config.Config) (blockchain.FpToken, blockchain.StableToken,
 blockchain.Erc721Token, error)
```

File: tools/erc721tester/assetcontract/assetcontract.go [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+	if err := erc721Token.Start(); err != nil {
+		return nil, err
+	}
+
+	return erc721Token, nil
+}
+
+// GenerateAssetID generates an asset ID
+func GenerateAssetID() string {
```
Comment (lizhefeng) [inline tools/erc721tester/assetcontract/assetcontract.go]: Delete this and reuse the one in executiontester/assetcontract/assetcontract.go

File: tools/erc721tester/assetcontract/assetcontract.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+		if len(id) >= 8 {
+			// attach 8-digit YYYYMMDD at front
+			t := time.Now().Format(time.RFC3339)
+			t = strings.Replace(t, "-", "", -1)
+			return t[:8] + id[:8]
+		}
+	}
+}
+
+func deployContract(code, endpoint string, args ...[]byte) (string, error) {
```
Comment (lizhefeng) [inline tools/erc721tester/assetcontract/assetcontract.go]: Delete this and reuse the one in executiontester/assetcontract/assetcontract.go

File: tools/erc721tester/blockchain/erc721.sol [hunk 16]
```diff
@@ -0,0 +1,967 @@
+pragma solidity 0.4.24;
```
Comment (lizhefeng) [inline tools/erc721tester/blockchain/erc721.sol]: You can move this file under executiontester/blockchain/

File: tools/erc721tester/blockchain/erc721_token.go [hunk 17]
```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2019 IoTeX
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package blockchain
```
Comment (lizhefeng) [inline tools/erc721tester/blockchain/erc721_token.go]: Move this file under executiontester/blockchain/

File: tools/erc721tester/blockchain/typedef.go [hunk 18]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) 2019 IoTeX
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package blockchain
```
Comment (lizhefeng) [inline tools/erc721tester/blockchain/typedef.go]: Move this file under executiontester/blockchain/

File: tools/erc721tester/erc721tester.go [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+	"github.com/iotexproject/iotex-core/config"
+	"github.com/iotexproject/iotex-core/pkg/keypair"
+	"github.com/iotexproject/iotex-core/pkg/log"
+	"github.com/iotexproject/iotex-core/pkg/probe"
+	"github.com/iotexproject/iotex-core/server/itx"
+	"github.com/iotexproject/iotex-core/testutil"
+	"github.com/iotexproject/iotex-core/tools/erc721tester/assetcontract"
+)
+
+func main() {
```
Comment (lizhefeng) [inline tools/erc721tester/erc721tester.go]: Merge this test into "executiontester.go", then the test structure looks like this:
1. start a single IoTeX server; (shared by erc721_token and fp_token);
2. create two accounts; (shared by erc721_token and fp_token);
3. create two (three if taking stable token into account) token contracts (prerequisite deployment should be done in function StartContracts, then invoke CreateToken respectively);
4. transfer fp token between the two accounts;
5. transfer erc721 token between the two account;
5. compare the actual sum of fp tokens with the amount of creation;
6. compare the actual sum of erc721 tokens with the amount of creation.

**Comment by lzxm160** [comment]
modified all as mentioned

File: .vscode/launch.json [hunk 20]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (lizhefeng) [inline .vscode/launch.json]: Please remove this file.

File: tools/executiontester/executiontester.go [hunk 21]
```diff
@@ -80,27 +83,45 @@ func main() {
...
+	if err != nil {
+		log.L().Fatal("Failed to transfer 1 token from creditor to debtor", zap.Error(err))
+	}
+
 	if _, err := fpToken.RiskLock(contractAddr, creditorAddr, creditorPriKey, risk); err != nil {
 		log.L().Fatal("Failed to transfer amount of risk from creditor to contract", zap.Error(err))
 	}
 
-	debtorBalance, err := fpToken.ReadValue(contractAddr, "70a08231", debtorAddr)
+	// Wait until transfer is successfully
```
Comment (lizhefeng) [inline tools/executiontester/executiontester.go]: Why do we need this wait until?

File: tools/executiontester/executiontester.go [hunk 22]
```diff
@@ -80,27 +83,45 @@ func main() {
...
 	}
-
+	// create fpToken with assetID
 	contractAddr, err := fpToken.TokenAddress(assetID)
 	if err != nil {
 		log.L().Fatal("Failed to get token contract address", zap.Error(err))
 	}
-
+	// create erc721 token with tokenID using assetID
+	if _, err := erc721Token.CreateToken(assetID, creditorAddr); err != nil {
```
Comment (lizhefeng) [inline tools/executiontester/executiontester.go]: Don't you need a new assetID for erc721Token?

File: .vscode/launch.json (hunk 20 above)
Comment (lzxm160) [reply .vscode/launch.json]: removed

File: tools/executiontester/executiontester.go (hunk 22 above)
Comment (lzxm160) [reply tools/executiontester/executiontester.go]: added a new id for erc721

**Comment by lzxm160** [comment]
> LGTM now. Just one more comment: I don't think we need the "wait until" inside the main function. Please remove it before merge. Thanks.

ok,but i don't understand,why it's necessary for erc721 only?

**Pull request merged**